Drivers/PriceVariantsPartDisplayDriver.cs

```diff
@@ -12,49 +15,66 @@ namespace OrchardCore.Commerce.Drivers
         }
 
-        public override IDisplayResult Edit(PriceVariantsPart pricePart)
+        public override IDisplayResult Edit(PriceVariantsPart part, BuildPartEditorContext context)
         {
-            return Initialize<PriceVariantsPartViewModel>("PriceVariantsPart_Edit", m => BuildViewModel(m, pricePart));
+            return Initialize<PriceVariantsPartViewModel>(GetEditorShapeType(context), m =>
+            {
+                BuildViewModel(m, part);
+                m.Currencies = _moneyService.Currencies;
+            });
         }
 
-        public override async Task<IDisplayResult> UpdateAsync(PriceVariantsPart model, IUpdateModel updater)
+        public override async Task<IDisplayResult> UpdateAsync(PriceVariantsPart part, IUpdateModel updater, UpdatePartEditorContext context)
         {
             var updateModel = new PriceVariantsPartViewModel();
-            await updater.TryUpdateModelAsync(updateModel, Prefix, t => t.VariantsValues);
-            await updater.TryUpdateModelAsync(updateModel, Prefix, t => t.VariantsCurrencies);
-            model.Variants = updateModel.VariantsValues.ToDictionary(x => x.Key, x => _moneyService.Create(x.Value, updateModel.VariantsCurrencies[x.Key]));
+            if (await updater.TryUpdateModelAsync(updateModel, Prefix, t => t.VariantsValues, t => t.VariantsCurrencies))
+            {
+                // Remove any content or the variants would be merged and not be cleared
+                part.Content.Variants.RemoveAll();
 
-            return Edit(model);
+                part.Variants = updateModel.VariantsValues
+                    .Where(x => x.Value.HasValue
+                        && updateModel.VariantsCurrencies?.ContainsKey(x.Key) == true
+                        && updateModel.VariantsCurrencies[x.Key] != Currency.UnspecifiedCurrency.CurrencyIsoCode)
+                    .ToDictionary(x => x.Key,
+                        x => _moneyService.Create(x.Value.Value, updateModel.VariantsCurrencies[x.Key]));
+            }
+
+            return Edit(part, context);
         }
 
         private Task BuildViewModel(PriceVariantsPartViewModel model, PriceVariantsPart part)
         {
             model.ContentItem = part.ContentItem;
-
             model.PriceVariantsPart = part;
-            model.Currencies = _moneyService.Currencies;
 
-            model.Variants = _priceVariantsService.GetPriceVariants(part.ContentItem);
-            model.VariantsValues = model.Variants.ToDictionary(x => x.Key, x => x.Value.Value);
-            model.VariantsCurrencies = model.Variants.ToDictionary(x => x.Key, x => x.Value.Currency.CurrencyIsoCode);
+            var allVariantsKeys = _predefinedValuesProductAttributeService.GetProductAttributesCombinations(part.ContentItem);
+            model.Variants = part.Variants;
+
+            model.VariantsValues = allVariantsKeys.ToDictionary(x => x,
+                x => model?.Variants?.ContainsKey(x) == true
```

Comment:
You can probably simplify this with a `TryGetValue`, something like `model?.Variants?.TryGetValue(x, out Amount amount) == true ? amount.Value : null` if the compiler lets you. Same thing for currencies. It's weird however that your view model would have two distinct collections for values and currencies. Seems like this could be simpler by using amounts as long as possible?

Reply:
we need to have nullable decimal, so best way to use two collections, as for me.